Zend/zend_enum.c

```diff
@@ -0,0 +1,287 @@
+/*
+   +----------------------------------------------------------------------+
+   | Zend Engine                                                          |
+   +----------------------------------------------------------------------+
+   | Copyright (c) Zend Technologies Ltd. (http://www.zend.com)           |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: Ilija Tovilo <ilutov@php.net>                               |
+   +----------------------------------------------------------------------+
+*/
+
+#include "zend.h"
+#include "zend_API.h"
+#include "zend_compile.h"
+#include "zend_interfaces_arginfo.h"
+
+#define ZEND_ENUM_PROPERTY_ERROR() \
+	zend_throw_error(NULL, "Enum properties are immutable")
+
+#define ZEND_ENUM_DISALLOW_MAGIC_METHOD(propertyName, methodName) \
+	do { \
+		if (ce->propertyName) { \
+			zend_error_noreturn(E_COMPILE_ERROR, "Enum may not include %s", methodName); \
+		} \
+	} while (0);
+
+static zend_object_handlers enum_handlers;
+
+zend_object *zend_enum_new(zval *result, zend_class_entry *ce, zval *case_name_zv, zval *scalar_zv)
+{
+	// Temporarily remove the ZEND_ACC_ENUM flag to allow instantiation
+	// FIXME: I'm sure we'll find a cleaner solution
+	ce->ce_flags &= ~ZEND_ACC_ENUM;
+	object_init_ex(result, ce);
+	ce->ce_flags |= ZEND_ACC_ENUM;
+
+	zend_object *zobj = Z_OBJ_P(result);
+	if (scalar_zv != NULL) {
+		zobj->handlers->write_property(zobj, ZSTR_KNOWN(ZEND_STR_VALUE), scalar_zv, NULL);
+	}
+
+	zobj->handlers = &enum_handlers;
+
+	return zobj;
+}
+
+static void zend_verify_enum_properties(zend_class_entry *ce)
+{
+	zend_property_info *property_info;
+
+	ZEND_HASH_FOREACH_PTR(&ce->properties_info, property_info) {
+		if (zend_string_equals_literal(property_info->name, "case")) {
```

Comment:
Is this and ZEND_STR_CASE and the unit tests of `->case` left over from the previous revision of this?

May be useful to change the unit tests of `->case` to `->anyPropName` for clarity

Reply:
Yes, it's actually already fixed in my other wip branch.